tutor/src/screens/scores-report/weights-ux.jsx

```diff
@@ -10,86 +10,151 @@ const DEFAULTS = {
+
+const MIN = 0;
+const MAX = 100;
+const RANGE = MAX - MIN;
+
+const weightToPercent = (weight) => ((weight || MIN) * RANGE );
+const percentToWeight = (percent) => ((percent || MIN) / RANGE );
+
+const weightsToPercents = (weights) => mapValues(WC, (c) => weightToPercent(weights[c]));
+const percentsToWeights = (percents) => mapValues(CW, (w) => percentToWeight(percents[w]));
+
 export default class ScoresReportWeightsUX {
 
   @observable homework_scores;
   @observable homework_progress;
   @observable reading_scores;
   @observable reading_progress;
 
-  @observable hasChanged = false;
+  @observable hasTouched = false;
   @observable isSetting = false;
 
   constructor(scoresUx) {
     this.scoresUx = scoresUx;
   }
 
+  @computed get course() {
+    return this.scoresUx.course;
+  }
+
+  @computed get isBusy() {
+    return this.course.api.isPending;
+  }
+
   @action.bound onSetClick() {
     const { course } = this;
     this.isSetting = true;
-    each(CW, (w, c) => {
-      this[w] = (course[c] || 0) * 100;
-    });
+    Object.assign(this, this.currentPercents);
   }
 
   @action.bound onCancelClick() {
     this.isSetting = false;
   }
 
   @action.bound onSaveWeights() {
+    const { course, currentPercents } = this;
+    Object.assign(course, this.nextWeights);
+    course
+      .save()
+      .then(() => {
+        return course.scores
+          .fetch();
+      })
+      .catch(() => {
+        // reset course weights to previous values
+        Object.assign(course, percentsToWeights(currentPercents));
+      })
+      .then(() => {
+        this.isSetting = false;
+      });
+  }
+
+  @action.bound setWeight(ev) {
+    this.hasTouched = true;
+    const weight = parseInt(ev.target.value);
+    if (inRange(weight, MIN, MAX + 1)) { // inRange is up to but not including end
+      this[ev.target.name] = weight;
+    }
+  }
+
+  @computed get savedCourseWeightsAsPercents() {
     const { course } = this;
-    each(CW, (w, c) => {
-      course[c] = this[w] / 100;
-    });
-    course.save().then(() => {
-      course.scores.fetch();
-      this.isSetting = false;
-    });
+    return weightsToPercents(course);
   }
 
-  @computed get course() {
-    return this.scoresUx.course;
+  @computed get uxPercentsAsCourseWeights() {
+    return percentsToWeights(this);
   }
 
-  @computed get isBusy() {
-    return this.course.api.isPending;
+  @computed get currentPercents() {
+    return this.savedCourseWeightsAsPercents;
   }
 
-  @action.bound setWeight(ev) {
-    this.hasChanged = true;
-    const weight = parseInt(ev.target.value);
-    if (inRange(weight, 0, 101)) { // inRange is up to but not including end
-      this[ev.target.name] = weight;
-    }
+  @computed get nextWeights() {
+    return this.uxPercentsAsCourseWeights;
+  }
+
+  @computed get values() {
+    return pick(this, SETTINGS);
+  }
+
+  @computed get total() {
+    return flow(
+      toArray,
+      sum,
+      Math.round,
+    )(this.values);
```

Comment:
`toArray` is fine, but you could call `this.values.peek()` to get a reference to the internal array that mobx uses.  And if you wanted to modify the array you can call `.slice()` to get a new array

Reply:
okay

Reply:
ah actually cant use `peek` here because `values` is a computed object